example_code/shmem_team_context.c

```diff
@@ -1,106 +1,90 @@
 #include <stdio.h>
 
-int isum, ival;
 
-int my_ctx_translate_pe(shmem_ctx_t src_ctx, int src_pe, shmem_ctx_t dest_ctx) {
-  if (src_ctx == SHMEM_CTX_INVALID) {
-    return -1;
-  }
-  if (dest_ctx == SHMEM_CTX_INVALID) {
-    return -1;
-  }
-
-  shmem_team_t src_team, dest_team;
-  shmem_ctx_get_team(src_ctx, &src_team);
-  shmem_ctx_get_team(dest_ctx, &dest_team);
-  return shmem_team_translate_pe(src_team, src_pe, dest_pe);
-}
-
-shmem_ctx_t my_team_create_ctx(shmem_team_t team) {
-  if (team == SHMEM_TEAM_INVALID) {
-    return SHMEM_CTX_INVALID;
-  }
+int main(void)
+{
+  static int            sum = 0, val_2, val_3;
+  shmem_team_t          team_2, team_3;
+  shmem_ctx_t           ctx_2, ctx_3;
+  shmem_team_config_t   conf;
 
-  shmem_ctx_t ctx;
-  if (shmem_team_create_ctx(team, 0, &ctx) != 0) {
-    fprintf(stderr, "Failed to create context for PE team\n");
-    return SHMEM_CTX_INVALID;
-  }
-  return ctx;
-}
+  shmem_init();
 
-void my_send_to_neighbor(shmem_ctx_t ctx, int *val) {
-  if (ctx == SHMEM_CTX_INVALID) {
-    fprintf(stderr, "Send to neighbor fail due to invalid context\n");
-    return;
-  }
+  int npes = shmem_n_pes();
+  int mype = shmem_my_pe();
+  conf.num_contexts = 1;
+  long cmask = SHMEM_TEAM_NUM_CONTEXTS;
 
-  shmem_team_t team;
-  shmem_ctx_get_team(ctx, &team);
-  int team_mype = shmem_team_my_pe(team);
-  int team_npes = shmem_team_n_pes(team);
-  int rpe = (team_mype + 1) % team_npes;
+  /* Create team_2 with PEs numbered 0, 2, 4, ... */
+  shmem_team_split_strided(SHMEM_TEAM_WORLD, 0, 2, (npes + 1) / 2, &conf, cmask, &team_2);
 
-  // put my pe number in the buffer on my right hand neighbor
-  shmem_ctx_int_put(ctx, val, &pe, 1, rpe);
-}
+  /* Create team_3 with PEs numbered 0, 3, 6, ... */
+  shmem_team_split_strided(SHMEM_TEAM_WORLD, 0, 3, (npes + 2) / 3, &conf, cmask, &team_3);
 
-int main() {
-  shmem_init();
+  /* Create a context on team_2. */
+  int ret = shmem_team_create_ctx(team_2, 0, &ctx_2);
 
-  int npes = shmem_n_pes();
-  isum = 0;
+  if (ret != 0) {
+      printf("%d: Error creating context ctx_2 (%d)\n", mype, ret);
+      shmem_global_exit(ret);
+  }
 
-  shmem_team_t team_2s, team_3s;
-  shmem_ctx_t ctx_2s, ctx_3s;
-  shmem_team_config_t conf = {.num_contexts = 1};
-  long cmask = SHMEM_TEAM_NUM_CONTEXTS;
+  /* Create a context on team_3. */
+  ret = shmem_team_create_ctx(team_3, 0, &ctx_3);
 
-  // Create team with PEs numbered 0, 2, 4, ...
-  shmem_team_spit_strided(SHMEM_TEAM_WORLD, 0, 2, npes / 2, &conf, cmask, &team_2s);
-  // Create team with PEs numbered 0, 3, 6, ...
-  shmem_team_split_strided(SHMEM_TEAM_WORLD, 0, 3, npes / 3, &conf, cmask, &team_3s);
+  if (ret != 0) {
+      printf("%d: Error creating context ctx_3 (%d)\n", mype, ret);
+      shmem_global_exit(ret);
+  }
 
-  ctx_2s = my_team_create_ctx(team_2s);
-  ctx_3s = my_team_create_ctx(team_3s);
+  /* Within each team, put my PE number to my neighbor in a ring-based manner. */
+  if (ctx_2 != SHMEM_CTX_INVALID) {
```

Comment:
It may not be possible to reach this point with one of the contexts set to `SHMEM_CTX_INVALID` since the above return code checks should have filtered those cases out.

Reply:
Right...

Sorry, this test still appears to be broken.  Should we remove the check on `shmem_team_create_ctx` (or alternatively, just print out the "non-members")?  In that case, we'd need to keep this check for `SHMEM_CTX_INVALID` here to avoid a put on an invalid PE, right?

Reply:
I would keep the check on context creation. If that is successful, there should be valid contexts at this point in the code. Assigning the default context on failure (as we do in some of the SOS unit tests) doesn't work in this case, since you need a specific team to be associated with the context.

Reply:
Ok, but I feel like we need to modify the condition in the check on context creation.  Perhaps:

```c
if (ret != 0 && team_2 != SHMEM_TEAM_INVALID) {
      printf("%d: Error creating context ctx_2 (%d)\n", mype, ret);
      shmem_global_exit(ret);
  }
...
```
Non-team member PEs (which we expect to have `ret > 0` here) can't call `shmem_global_exit`, otherwise the test will never get past these checks.

Does this work, or am I missing something?

Reply:
Context creation can fail even with team_2 is valid. This change can allow `ctx2 == SHMEM_CTX_INVALID` to make it past this check.

Reply:
You could make it `ret != 0 || team_2 != SHMEM_TEAM_INVALID` but the latter check is redundant since the invalid team is required to cause ret to be nonzero.

Reply:
I do agree context creation can fail even when passed a valid team...

But doesn't this program _expect_ some PEs (PE 1 in particular) to end up with a  `SHMEM_TEAM_INVALID` team as well as a `SHMEM_CTX_INVALID` context?  If PE 1 calls exit, this example won't work with more than 1 PE.  

Perhaps the mistake is that `shmem_team_ctx_create` returns _nonzero_ when passed a `SHMEM_TEAM_INVALID` team...  If `ctx_create` returned zero when passed the invalid team, I don't think we'd have this funny business.

Reply:
Oh, you're right. `&&` looks right